plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/SortByInputModelProcessor.java

```diff
@@ -0,0 +1,166 @@
+        int layerIndex = 0;
+        for (Layer layer : graph) {
+            final int previousLayerIndex = layerIndex == 0 ? 0 : layerIndex - 1;
+            Layer previousLayer = graph.getLayers().get(previousLayerIndex);
+            for (LNode node : layer.getNodes()) {
+                Collections.sort(node.getPorts(), (p1, p2) -> {
+                    // Sort incoming edges by sorting their ports by the order of the nodes they connect to.
+                    if (p1.getOutgoingEdges().isEmpty() && p2.getOutgoingEdges().isEmpty()) {
+                        LNode p1Node = p1.getIncomingEdges().get(0).getSource().getNode();
+                        LNode p2Node = p2.getIncomingEdges().get(0).getSource().getNode();
+                        if (p1Node.equals(p2Node)) {
+                            // In this case both incoming edges must have a model order set. Check it.
+                            return Integer.compare(
+                                    p1.getIncomingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER),
+                                    p2.getIncomingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER));
+                        }
+                        for (LNode previousNode : previousLayer) {
+                            if (previousNode.equals(p1Node)) {
+                                return 1;
+                            }
+                            if (previousNode.equals(p2Node)) {
+                                return -1;
+                            }
+                        }
+                    }
+
+                    // Sort outgoing edges by sorting their ports.
+                    if (p1.getIncomingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return Integer.compare(
+                                p1.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER),
+                                p2.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER));
+                    }
+                    // Sort outgoing ports before incoming ports.
+                    if (p1.getOutgoingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return 1;
+                    } else {
+                        return -1;
+                    }
+                });
+            }
+            // Sort nodes.
+            Collections.sort(layer.getNodes(), (n1, n2) -> {
+                // If no model order is set, the one node is a dummy node and the nodes should be ordered
+                // by the connected edges.
+                if (!n1.hasProperty(InternalProperties.MODEL_ORDER)
+                        || !n2.hasProperty(InternalProperties.MODEL_ORDER)) {
+                    // In this case the order of the connected nodes in the previous layer should be respected
+           
+                    LPort p1 = null;
```

Comment:
`n1.getPorts().stream().filter(p -> !p.incoming.empty).findFirst()`?

Reply:
This seems reasonable.